File: QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
 package modeloqytetet;

import java.util.ArrayList;

/**
 *
 * @author david
 */
public class Jugador {
    private boolean encarcelado;
    private String nombre;
    private int saldo;
    private ArrayList<TituloPropiedad> propiedades;
    private Casilla casillaActual;
    private Sorpresa cartaLibertad;
    static int factorEspeculador = 1;
    
    
    public Jugador(String nombre) {
        encarcelado = false;
        this.nombre = nombre;
        saldo = 7500;
        cartaLibertad = null;
        casillaActual = null;
        propiedades = new ArrayList();
    }
    
    protected Jugador(Jugador jugador){
        encarcelado = jugador.getEncarcelado();
        nombre = jugador.getNombre();
        saldo = jugador.getSaldo();
        propiedades = jugador.propiedades;
        casillaActual = jugador.getCasillaActual();
        cartaLibertad = jugador.cartaLibertad;
        
        for(TituloPropiedad miTitulo : propiedades)
            miTitulo.setPropietario(this);
    }
    
    public int getFactorEspeculador() {
        return factorEspeculador;
    }

    public String getNombre() {
        return nombre;
    }
    
    public Casilla getCasillaActual() {
        return casillaActual;
    }
    
    public boolean getEncarcelado() {
        return encarcelado;
    }
    
    public int getSaldo() {
        return saldo;
    }
    
    public boolean tengoPropiedades() {
        return !(propiedades.isEmpty());
    }   
    
    boolean actualizarPosicion(Casilla casilla) { 
        if(casilla.getNumeroCasilla() < casillaActual.getNumeroCasilla())
            modificarSaldo(Qytetet.PRECIO_SALIDA);
        
        boolean tengoPropietario = false;
        setCasillaActual(casilla);        
        
        if(casilla.soyEdificable()) {
            tengoPropietario = ((Calle)casilla).tengoPropietario();
            if(tengoPropietario) {
                encarcelado = ((Calle)casilla).propietarioEncarcelado();
                if(!encarcelado) {
                    int costeAlquiler = ((Calle)casilla).cobrarAlquiler();
                    modificarSaldo(-costeAlquiler);
                }
            }
        }
        else {
            if(casilla.getTipo() == TipoCasilla.IMPUESTO){
                int coste = casilla.getCoste();
                pagarImpuestos(coste);
                System.out.println("\n¡IMPUESTO! Pierdes "+(-coste));
            }
        }
        return tengoPropietario;
    }

    
    boolean comprarTitulo() {
        boolean puedoComprar = false;
        
        if(casillaActual.soyEdificable()) {
            boolean tengoPropietario = ((Calle)casillaActual).tengoPropietario();
            if(!tengoPropietario) {
                int costeCompra = ((Calle)casillaActual).getCoste();
                if(costeCompra <= saldo) {
                    TituloPropiedad titulo = ((Calle)casillaActual).asignarPropietario(this);
                    titulo.setCasilla(((Calle)casillaActual));
                    propiedades.add(titulo);
                    modificarSaldo(-costeCompra);
                    puedoComprar = true;
                }
            }
        }
        
        return puedoComprar;
    }
    
    Sorpresa devolverCartaLibertad() {
        Sorpresa cartaDevuelta = cartaLibertad;
        cartaLibertad = null;
        
        return cartaDevuelta;
    }
    
    void irACarcel(Casilla casilla) {
        setCasillaActual(casilla);
        setEncarcelado(true);
    }
    
    void modificarSaldo(int cantidad) {
        saldo += cantidad;
    }
    
    int obtenerCapital() {
        int capital = saldo;
        
        for (TituloPropiedad misTitulos : propiedades) {
            capital += misTitulos.getCasilla().getCoste() + (misTitulos.getCasilla().getNumCasas() + 
            misTitulos.getCasilla().getNumHoteles()) * misTitulos.getPrecioEdificar();
            
            if (misTitulos.getHipotecada())
                capital -= misTitulos.getHipotecaBase();
        }
        
        return capital;
    }
    
    ArrayList<TituloPropiedad> obtenerPropiedadesHipotecadas(boolean hipotecada) {
        ArrayList<TituloPropiedad> propiedadesHipotecasdas = new ArrayList();
        
        for (TituloPropiedad miTitulo : propiedades)
            if(miTitulo.getHipotecada() == hipotecada)
                propiedadesHipotecasdas.add(miTitulo);
        
        return propiedadesHipotecasdas;
    }    
    
    void pagarCobrarPorCasaYHotel(int cantidad) {
        int numeroTotal = cuantasCasasHotelesTengo();
        modificarSaldo(numeroTotal*cantidad);
    }
    
    boolean pagarLibertad(int cantidad) {
        if(tengoSaldo(cantidad))
            modificarSaldo(-cantidad);
        
        return tengoSaldo(0);
    }
    
    boolean puedoEdificarCasa(Calle calle) {
        boolean puedoEdificar = false;
        boolean esMia = esDeMiPropiedad(calle);
        
        if (esMia) {
            int costeEdificarCasa = calle.getPrecioEdificar();
            boolean tengoSaldo = tengoSaldo(costeEdificarCasa);
            puedoEdificar = esMia && tengoSaldo;
        }
        
        return puedoEdificar;
    }

    boolean puedoEdificarHotel(Calle calle) {
        boolean puedoEdificar = false;
        boolean esMia = esDeMiPropiedad(calle);
        
        if (esMia) {
            int costeEdificarHotel = calle.getPrecioEdificar();
            boolean tengoSaldo = tengoSaldo(costeEdificarHotel);
            puedoEdificar = esMia && tengoSaldo;
        }
        
        return puedoEdificar;
    }
    
    boolean puedoHipotecar(Calle calle) { 
        return esDeMiPropiedad(calle);
    }

    boolean puedoPagarHipoteca(Calle calle) {
        return saldo >= calle.cancelarHipoteca();
    }

    boolean puedoVenderPropiedad(Calle calle) {
        return esDeMiPropiedad(calle) && !calle.estaHipotecada();
    }
    
    void setCartaLibertad(Sorpresa carta) {
        cartaLibertad = carta;
    }

    void setCasillaActual(Casilla casilla) {
        casillaActual = casilla;
    }

    void setEncarcelado(boolean encarcelado) {
        this.encarcelado = encarcelado;
    }

    
    boolean tengoCartaLibertad() {
        return cartaLibertad != null;
    }

    void venderPropiedad(Calle calle) {
        int precioVenta = calle.venderTitulo();
        modificarSaldo(precioVenta);
        eliminarDeMisPropiedades(calle);
    }
    
    protected void pagarImpuestos(int cantidad) {
        modificarSaldo(-cantidad);
    }
    
    
    protected Especulador convertirme(int fianza) {
        Especulador especulador = new Especulador (this, fianza);
        
        return especulador;
    }
    
    private int cuantasCasasHotelesTengo() {
        int numCasasHoteles = 0;
        for (TituloPropiedad miTitulo : propiedades)
            numCasasHoteles += miTitulo.getCasilla().getNumHoteles() +
                    miTitulo.getCasilla().getNumCasas();
            
        return numCasasHoteles;
    }
    
    private void eliminarDeMisPropiedades(Calle calle) {
        TituloPropiedad miTitulo = null;
        for (TituloPropiedad titulo : propiedades)
            if (calle == titulo.getCasilla())
                miTitulo = titulo;
        if (miTitulo != null)
            propiedades.remove(miTitulo);
    }

    private boolean esDeMiPropiedad(Calle calle) {
        boolean mePertenece = false;
        for (TituloPropiedad misTitulos : propiedades)
            if (calle == misTitulos.getCasilla())
                mePertenece = true;
        
        return mePertenece;    
    }

    private boolean tengoSaldo(int cantidad) {
        return saldo >= cantidad;
    }
    
    @Override
    public String toString() {
        return getNombre() + "\nSaldo: " + getSaldo();
    }

}

File: QytetetRuby/lib/jugador.rb
# encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


module ModeloQytetet

  class Jugador    
    attr_accessor :encarcelado, :nombre, :saldo, :propiedades, 
      :casilla_actual, :carta_libertad, :factor_especulador    

    def initialize(nombre)
      @nombre = nombre
      @saldo = 7500
      @encarcelado = false
      @propiedades = []
      @casilla_actual = nil
      @carta_libertad = nil
      @factor_especulador = 1
    
    end
    
    def copia_jugador(jugador)
      @nombre = jugador.nombre
      @saldo = jugador.saldo
      @encarcelado = jugador.encarcelado
      @propiedades = jugador.propiedades
      @casilla_actual = jugador.casilla_actual
      @carta_libertad = jugador.carta_libertad     

      if @propiedades != nil
        for mi_titulo in @propiedades
          mi_titulo.propietario = self
        end
      end
    
    #  @factor_especulador = jugador.factor_especulador
    end
    
    def actualizar_posicion(casilla)
      if(casilla.numero_casilla < @casilla_actual.numero_casilla)
        modificar_saldo(Qytetet.saldo_salida)
      end
      @casilla_actual = casilla
      tengo_propietaro = false
      
      if casilla.soy_edificable
        tengo_propietaro = casilla.tengo_propietario
        if tengo_propietaro
          @encarcelado = casilla.propietario_encarcelado
          if !@encarcelado
            coste_alquiler = casilla.cobrar_alquiler
            modificar_saldo(-coste_alquiler)
          end
        end
      elsif casilla.tipo == TipoCasilla::IMPUESTO
          coste = casilla.coste
          pagar_impuestos(coste)
          puts "\n¡IMPUESTO! Pierdes - #{coste}"
      end
      
      return tengo_propietaro
    end
    
    def comprar_titulo
      puedo_comprar = false
      
      if @casilla_actual.soy_edificable
        tengo_propietario = @casilla_actual.tengo_propietario
        if !tengo_propietario
          coste_compra = @casilla_actual.coste
          if coste_compra <= @saldo
            titulo = @casilla_actual.asignar_propietario(self)
            titulo.casilla = @casilla_actual
            @propiedades<< titulo
            modificar_saldo(-coste_compra)
            puedo_comprar = true
          end
        end
      end
      
      return puedo_comprar
    end
    
    def devolver_carta_libertad
      carta_devuelta = @carta_libertad
      @carta_libertad = nil
      return carta_devuelta
    end
    
    def ir_a_carcel(casilla)
      @casilla_actual = casilla
      @encarcelado = true
    end
    
    def modificar_saldo(cantidad)
      @saldo += cantidad
    end
    
    def obtener_capital
      capital = @saldo
      
      for titulos in @propiedades
        capital += titulos.casilla.coste + 
          (titulos.casilla.num_casas + titulos.casilla.num_hoteles)*(titulos.precio_edificar)
        if titulos.hipotecada
          capital -= titulos.hipoteca_base
        end
      end
      
      return capital
    end
    
    def obtener_propiedades_hipotecadas(hipotecada)
      propiedades_hipotecadas = []
      
      for titulo in @propiedades
        if titulo.hipotecada == hipotecada
          propiedades_hipotecadas<< titulo
        end
      end
      
      return propiedades_hipotecadas
    end
    
    def pagar_cobrar_por_casa_y_hotel(cantidad)
      numero_total = cuantas_casas_hoteles_tengo
      modificar_saldo(numero_total*cantidad)
    end
    
    def pagar_impuestos(cantidad)
      modificar_saldo(-cantidad)
    end
    
    def pagar_libertad(cantidad)
      if tengo_saldo(cantidad)
        modificar_saldo(-cantidad)
      end
      
      return tengo_saldo(cantidad)
    end
    
    def puedo_edificar_casa(casilla)
      puedo_edificar = false
      es_mia = es_de_mi_propiedad(casilla)
      
      if es_mia
        coste_edificar_casa = casilla.precio_edificar
        tengo_saldo = tengo_saldo(coste_edificar_casa)
        puedo_edificar = es_mia && tengo_saldo
      end
      
      return puedo_edificar
    end
    
    def puedo_edificar_hotel(casilla)
      puedo_edificar = false
      es_mia = es_de_mi_propiedad(casilla)
      
      if es_mia
        coste_edificar_hotel = casilla.precio_edificar
        tengo_saldo = tengo_saldo(coste_edificar_hotel)
        puedo_edificar = es_mia && tengo_saldo
      end
      
      return puedo_edificar
    end
    
    def puedo_hipotecar(casilla)
      return es_de_mi_propiedad(casilla)
    end
    
    def puedo_vender_propiedad(casilla)
      return es_de_mi_propiedad(casilla) && !casilla.esta_hipotecada
    end
    
    def puedo_pagar_hipoteca(casilla)
      return @saldo >= casilla.cancelar_hipoteca      
    end
    
    def tengo_carta_libertad
      return @carta_libertad != nil
    end
    
    def vender_propiedad(casilla)
      precio_venta = casilla.vender_titulo
      modificar_saldo(precio_venta)
      eliminar_de_mis_propiedades(casilla)
    end
    
    def tengo_propiedades
      return !(@propiedades.empty?)
    end
    
    def convertirme(fianza)
      especulador = Especulador.new(self,fianza)
      return especulador
    end
    
    def to_s
      "\nNombre: #{@nombre} \nSaldo: #{@saldo}"
    end
    
    private
    def cuantas_casas_hoteles_tengo
      num_casas_hoteles = 0
      for titulo in @propiedades
        num_casas_hoteles += titulo.casilla.num_hoteles + titulo.casilla.num_casas
      end
      
      return num_casas_hoteles      
    end
    
    def eliminar_de_mis_propiedades(casilla)
      @propiedades.delete(casilla.titulo)
    end
    
    def es_de_mi_propiedad(casilla)
      me_pertenece = false;
      for titulo in @propiedades
        if casilla == titulo.casilla
          me_pertenece = true
        end
      end
      
      return me_pertenece
    end
    
    def tengo_saldo(cantidad)
      return @saldo >= cantidad
    end
    
  end
  
end
